### PR TITLE
feat: add source mapping for calibration expansion

### LINKED ISSUE #366: Expand calibrations with source mapping
Currently, calibration expansion happens somewhat opaquely: a `Vec<Instruction>` goes in, and another `Vec<Instruction>` comes out. In complex programs, it can be hard for a user to discern which post-calibration instructions originated from which pre-calibration instructions.

This can be fixed with a source map. Perhaps it can be structured like this:

```
struct SourceMap<SourceIndex, TargetIndex> {
    entries: Vec<SourceMapEntry<SourceIndex, TargetIndex>>
}

struct SourceMapEntry<Label, SourceIndex, TargetIndex> {
    /// A description of the mapping, such the calibration used to perform the expansion
    label: Label,

    source_location: Vec<SourceIndex>,
    target_location: Vec<TargetIndex>,
}
```

these are generic over `*Index` in order to be re-used for other mappings, such as in parsing a program (mapping `nom::locate` information such as line and column span). `Label`, likewise, would be a different type in these situations:

* For calibration expansion, may be the calibration used to perform the expansion. Users like @bramathon have often wondered about which DEFCAL was actually being used, and this could help make that explicit. We'd have to figure out how this works for nested expansions.
* For parsing, might just be `()`, I don't see a need for a label

This type would be made available for other libraries to use as well, such as compilers which consume Quil.

Finally, `SourceMap`s should be able to be folded, such that a program can undergo multiple transformations in series and emit a single mapping from the original text down to the final artifact.

### PR DESCRIPTION
**Pull request by kalzoo**
This PR allows the (Rust package) user to do the following:

* Expand a single instruction in a self-descriptive, recursive way. That is, expand an instruction using its calibrations, and alongside the newly expanded instructions, receive an object describing the calibration decisions made and which of the new instructions are a result of which calibration decisions.
* Expand an entire program in the same way, receiving not only the expanded program but a source mapping explaining how every instruction in the new program was derived from the first one: either it was expanded (and it will show the expanded instruction as well as the DEFCAL [MEASURE]s used to expand it) or it was not (and it will yield the instruction index in both source and target program which contains it)

Reviewer:

* Please consider the two comments suggesting a breaking change to the `MeasureCalibrationDefinition` and `Calibration` structs. This would make them somewhat more similar to DEFFRAME in that they would have identifiers that could be used to index them without respect to their instruction bodies. We could also take the chance to rename `Calibration` to `CalibrationDefinition` for consistency. 
  * **I have made this change, but in a way that should not break some python consumers.**
  * All of the original `@property`s are there, so this will only break users who construct and modify calibrations and measure calibrations. I could add additional setters to help migrate, but there's only one `__new__` and it doesn't feel right to leave those behind.
* Please take a look at the new tests and consider the SourceMap format. Do you see any shortcomings?
* Please look at the `CalibrationSet#_expand` method and those nearby.
  * I'd like to have this only construct a `CalibrationExpansion` when that is returned to the user; I first tried an `expansion: Option<&mut CalibrationExpansion>` argument, but that doesn't work well here because that doesn't have a sensible default.
  * So instead I went with `build_source_map` which skips _most_ of the work involved in calibration expansion source mapping while still technically constructing one which is then discarded prior to return from `CalibrationSet.expand`.
  * Can you suggest a more elegant solution? I want to share logic between expand-with-source-map and expand-without-source-map while not doing any source mapping work if one isn't going to be returned.

TODOs:

- [x] Fix the bug in program calibration expansion source mapping
- [x] Restore the "fast path" to instruction calibration expansion when no source map is desired. In the current implementation, it's built in all cases and then thrown away in some in order to maintain the existing API.
- [x] (maybe, at request) An ergonomic reverse-mapping to allow looking up the source instruction that resulted in a given target instruction. That can always be built by the user, but it might be nice to have one here since the source mapping structure takes a moment to grok even for an experienced Quil dev.
- [ ] (maybe) implement source mapping for parsing. This could be done on an experimental branch just to validate that this PR has the right API to support that.

Closes #366

### CODE REVIEW

**Comment by mhodson-rigetti** [comment]
As much as I can read Rust, this is making sense to me. On the first point about a lurking bug, I wonder if `impl SourceMapRange for CalibrationExpansion` `contains()` is incomplete as containment is only by the index being valid, not that it also came from the same calibration? On the third point about reverse-mapping help, if I could see the pattern for this implemented using only the Python bindings as a recipe I think I'd be fine, but if doing that means you might as well codify on the Rust side, happy for that also!

**Comment by kalzoo** [comment]
> As much as I can read Rust, this is making sense to me. On the first point about a lurking bug, I wonder if `impl SourceMapRange for CalibrationExpansion` `contains()` is incomplete as containment is only by the index being valid, not that it also came from the same calibration? On the third point about reverse-mapping help, if I could see the pattern for this implemented using only the Python bindings as a recipe I think I'd be fine, but if doing that means you might as well codify on the Rust side, happy for that also!

Hey, thanks @mhodson-rigetti . I've added python bindings and docs but no example (yet) - I'll have to think about how best to expose those direct mappings from source to target. Right now (following our chat a week ago) I made the structure recursive to model the recursive nature of calibration expansion, and that's the least intuitive part in practical use, because each level of expansion has its own index basis. There's an example describing that problem in the python docstrings now.

You were right about `SourceMapRange` for exactly that reason - two ranges may not share the same basis. I just removed `SourceMapRange` altogether in this iteration, I don't think this functionality needs to be so general-purpose just yet.

The lingering bug though was actually fixed in [this commit](https://github.com/rigetti/quil-rs/pull/370/commits/bf246b3cce0e9ea5da107638461688d7a756b641) and was something prompted by our chat a week ago - certain instructions being yoinked out of the instruction body and thus (necessarily) out the source map.

File: quil-rs/src/instruction/calibration.rs
Comment (kalzoo) [inline quil-rs/src/instruction/calibration.rs:40]: From @MarquessV : this should not be needed anymore if the `CalibrationIdentifier` can serve as the signature

File: quil-rs/src/instruction/calibration.rs
Comment (kalzoo) [inline quil-rs/src/instruction/calibration.rs:22]: @MarquessV is open to renaming and breaking this, as long as this PR is already breaking

File: quil-rs/src/program/mod.rs
Comment (kalzoo) [inline quil-rs/src/program/mod.rs:331]: This feels very clumsy and likely inefficient, but no more clever way occurred to me. Suggestions welcome.

Basically, we determine whether an instruction is a "header" or "body" instruction by whether or not it made the instruction body longer, and then based on that we excise it from the source mapping so that the `target_index` ranges stay accurate and don't map to a missing instruction

File: quil-py/src/program/source_map.rs
Comment (MarquessV) [inline quil-py/src/program/source_map.rs:29]: I'm not against this implementation, but an option is to return a Python range. That might be slightly more Pythonic:

```rs
pub fn range(&self) -> (usize, usize) {
    let range = py.import("builtins")?.get("range")?;
    let args = PyTuple::new(py, &[self.as_inner().range().start, self.as_inner().range().end]);
    range.call1(args)
}
```
Return type aside, I wonder if it's possible to only consume the inner range a single time, rather that once for the start and once for the end.

File: quil-py/src/program/source_map.rs
Comment (MarquessV) [inline quil-py/src/program/source_map.rs:40]: Any Python type that can derive PartialEq should also use the `impl_eq!` macro so that Python can leverage it as well.

File: quil-rs/src/program/calibration.rs
Comment (MarquessV) [inline quil-rs/src/program/calibration.rs:508]: Would it make sense to throw this in an `is_match(&self, gate: &Gate)` method on CalibrationIdentifier?

File: quil-rs/src/program/mod.rs
Comment (MarquessV) [reply quil-rs/src/program/mod.rs:331]: I don't necessarily have a better idea for this, but an observation that the distinction between "header" and "body" instruction comes up enough that maybe we should have a more concrete definition of it in the library? Like a method in `Instruction` maybe? I know it's not a part of the spec, but it is a practical one that this library makes decisions around.

**Review by antalsz — CHANGES_REQUESTED** [review]
Looks overall good to me, but I think some details need to be hashed out before we merge this

File: quil-rs/src/program/calibration.rs
Comment (kalzoo) [reply quil-rs/src/program/calibration.rs:508]: yep, good thought, done

File: quil-py/src/program/source_map.rs
Comment (kalzoo) [reply quil-py/src/program/source_map.rs:29]: sure - TIL this trick with pyo3. It's compiling but I don't think I did it right; will fix it up and test on the next pass.

File: quil-py/src/program/source_map.rs
Comment (kalzoo) [reply quil-py/src/program/source_map.rs:40]: good call - added some, will sweep for rest on next pass

File: quil-rs/src/program/mod.rs
Comment (kalzoo) [reply quil-rs/src/program/mod.rs:331]: > if the calibration expansion can check if something is a "header" instruction, then it can refrain from adding it to the expanded source, and then we don't need to do this fixup.

I don't see how this is possible. When the instruction is expanded, it does not have the scope of a program, and so it is just expanded into a `Vec<Instruction>`. However, when we add that Vec to a program here, then some of its members are "hoisted" into data structures like memory region declarations and calibration declarations and such. Expanding the calibration in such a way that "predicted" how the program would handle it, without the use of an actual program, is something that could fall out of sync and requires extra care and maintenance.

File: quil-rs/src/program/source_map.rs
Comment (kalzoo) [reply quil-rs/src/program/source_map.rs:12]: yeah, I've had that in mind from the start, but the first attempt was clumsy given the generics involved. I've added one now, though - good interface, although perhaps less efficient than it could be.

> (maybe, at request) An ergonomic reverse-mapping to allow looking up the source instruction that resulted in a given target instruction. That can always be built by the user, but it might be nice to have one here since the source mapping structure takes a moment to grok even for an experienced Quil dev.

File: quil-py/src/program/source_map.rs
Comment (kalzoo) [reply quil-py/src/program/source_map.rs]: got this fixed up and under test - looking good. please lmk what you think

File: quil-rs/src/program/calibration.rs
Comment (kalzoo) [reply quil-rs/src/program/calibration.rs]: how would we maintain that as an invariant, if one expands to something that is all headers?

**Review by kalzoo — COMMENTED** [review]
Thanks for the review, @antalsz

File: quil-rs/src/instruction/calibration.rs
Comment (MarquessV) [reply quil-rs/src/instruction/calibration.rs:40]: That is true, I don't know of a good way to make the generics work with just a `CalibrationIdentifier` struct. I agree that it makes sense to formalize the concept of a calibration identifier w/ a struct and defer to it to implement the signature logic. They serve different purposes, so it's not entirely redundant, and the core logic doesn't need to be duplicated between either, so I'm good with this.

File: quil-py/src/program/source_map.rs
Comment (antalsz) [reply quil-py/src/program/source_map.rs]: Looks good to me!

File: quil-rs/src/instruction/calibration.rs
Comment (antalsz) [inline quil-rs/src/instruction/calibration.rs:156]: This completely inverts the meaning of the function but nothing else seems to have changed?

File: quil-rs/src/program/calibration.rs
Comment (antalsz) [inline quil-rs/src/program/calibration.rs:93]:
```suggestion
        // Adjust the start of the range if the target index is before the range
```

File: quil-rs/src/instruction/calibration.rs
Comment (antalsz) [reply quil-rs/src/instruction/calibration.rs:78]: Ahh, that makes sense – I was preemptively normalizing.

File: quil-rs/src/program/calibration.rs
Comment (antalsz) [reply quil-rs/src/program/calibration.rs]: Since I did this review I have come to understand that headers move 👍

File: quil-rs/src/program/calibration.rs
Comment (antalsz) [reply quil-rs/src/program/calibration.rs:465]: I have read the code again and figured out what I was missing, which is that sometimes the ranges aren't created at index `0`.  Sorry for the miss

File: quil-rs/src/program/mod.rs
Comment (antalsz) [reply quil-rs/src/program/mod.rs:335]: It's statically knowable which instructions get hoisted and which ones don't: `Instruction::CalibrationDefinition` does, `Instruction::Gate` doesn't, etc.  Expansion could return *two* vectors: `Vec<HoistedInstruction>` and `Vec<ComputationInstruction>` (new types optional, but perhaps clearer).  Then anything working on the results of expansion wouldn't have to do its own classification.

**Review by antalsz — COMMENTED** [review]
These changes look good, thanks!  I have a few open threads that weren't addressed, but none I'd consider blockers.  However, the discussion around header/body instruction distinction should wrap up before we merge this.

File: quil-rs/src/instruction/calibration.rs
Comment (kalzoo) [reply quil-rs/src/instruction/calibration.rs:156]: Yeah, it was just fixing a typo made within this MR when I moved this from its [previous home](https://github.com/rigetti/quil-rs/pull/370/files#diff-1816364e525c509ac2bb6d523144a0f2c0fc33697abf580e045bbb6e25afa74bL354)

Comment (antalsz) [reply quil-rs/src/instruction/calibration.rs:156]: Whoops, I should've caught that during my last review, sorry!

**Review by antalsz — APPROVED** [review]
As I said in my last review – now that I understand the negation change, I would not consider any of my remaining comments blocking.  Approved!

File: quil-py/quil/program/__init__.pyi
Comment (antalsz) [inline quil-py/quil/program/__init__.pyi:41]: I think this ought to have an assert.  That raises questions of the precise whitespace, I know.  Could we parse them both to Quil and assert the parsed programs are equal?

File: quil-py/quil/program/__init__.pyi
Comment (antalsz) [inline quil-py/quil/program/__init__.pyi:44]: I like this section a lot, but I found the specific comments a bit confusing.  Perhaps we could revise this first comment to something like

"…can interrogate the expansion source mapping.  For instance, The X at index 0 should have been replaced with a Z at index 0.  Here's how we could confirm that: First, we…".

As it stands, the comment on line 46 made me think that we were also going to check the Y at index 1.  (I think adding that would probably be redundant.)

Another possibility would be to drop the ellipses on lines 51 and 63; they're not following on from a previous sentence.

File: quil-py/quil/program/__init__.pyi
Comment (antalsz) [inline quil-py/quil/program/__init__.pyi:9]: Markdown code blocks with ` ``` ` don't need to have their contents indented, so we could dedent what follows.

File: quil-py/quil/program/__init__.pyi
Comment (antalsz) [inline quil-py/quil/program/__init__.pyi:8]: Is this snippet run and tested?  It would be nice if it were but I don't know enough about Python to tell if that's happening/how easy it would be to have that happen.

**Review by antalsz — COMMENTED** [review]
Reviewed 5d31908ca31f86e87c0baf1e45e57b21b2c0e822; a couple minor thoughts about the explanation, but the overall concept is great.

File: quil-py/quil/program/__init__.pyi
Comment (kalzoo) [reply quil-py/quil/program/__init__.pyi:8]: yes, it's a test case

File: quil-py/quil/program/__init__.pyi
Comment (kalzoo) [reply quil-py/quil/program/__init__.pyi:9]: ah, of course. will fix

Comment (kalzoo) [reply quil-py/quil/program/__init__.pyi]: fixed

File: quil-py/quil/program/__init__.pyi
Comment (kalzoo) [reply quil-py/quil/program/__init__.pyi]: the test did, I left it out here for readability but have restored it now

File: quil-py/quil/program/__init__.pyi
Comment (antalsz) [reply quil-py/quil/program/__init__.pyi:8]: Can we add a pointer to the test case it corresponds to?

File: quil-py/quil/program/__init__.pyi
Comment (kalzoo) [reply quil-py/quil/program/__init__.pyi]: yep, fixed

**Comment by bramathon** [comment]
Just wanted to check on the status of this - it's important for pulse plotting to work.

**Comment by bramathon** [comment]
I'm a little surprised by the extent of the changes in this PR. I was expecting something more along the lines of this:

```python
from quil.program import Program
from typing import Tuple, List

def expand_with_source_mapping(program: Program) -> Tuple[Program, List[int]]:
    """
    Expand any instructions in the program which have a matching calibration, leaving the others
    unchanged. 

    :param program: A quil.Program.
    :return: A quil.Program with the instructions expanded and a source map.
    The source map is a list with the length of the expanded instructions, and indicates the index
    of the logical instruction in the original program from which the expanded instruction originated.
    """
    instructions = program.body_instructions
    calibrations = program.calibrations
    expanded_program = program.clone_without_body_instructions()

    source_map = []
    expanded_instructions = []
    for logical_index, inst in enumerate(instructions):
        expanded_instruction = calibrations.expand(inst, [])
        source_map += [logical_index]*len(expanded_instruction)
        expanded_instructions += expanded_instruction
    expanded_program.add_instructions(expanded_instructions)
    return expanded_program, source_map

expanded_program, source_map = expand_with_source_mapping(quil_program)
default_expanded_program = quil_program.expand_calibrations()
assert expanded_program == default_expanded_program
```

**Comment by kalzoo** [comment]
> I'm a little surprised by the extent of the changes in this PR. I was expecting something more along the lines of this:
> 
> ```python
> from quil.program import Program
> from typing import Tuple, List
> 
> def expand_with_source_mapping(program: Program) -> Tuple[Program, List[int]]:
>     """
>     Expand any instructions in the program which have a matching calibration, leaving the others
>     unchanged. 
> 
>     :param program: A quil.Program.
>     :return: A quil.Program with the instructions expanded and a source map.
>     The source map is a list with the length of the expanded instructions, and indicates the index
>     of the logical instruction in the original program from which the expanded instruction originated.
>     """
>     instructions = program.body_instructions
>     calibrations = program.calibrations
>     expanded_program = program.clone_without_body_instructions()
> 
>     source_map = []
>     expanded_instructions = []
>     for logical_index, inst in enumerate(instructions):
>         expanded_instruction = calibrations.expand(inst, [])
>         source_map += [logical_index]*len(expanded_instruction)
>         expanded_instructions += expanded_instruction
>     expanded_program.add_instructions(expanded_instructions)
>     return expanded_program, source_map
> 
> expanded_program, source_map = expand_with_source_mapping(quil_program)
> default_expanded_program = quil_program.expand_calibrations()
> assert expanded_program == default_expanded_program
> ```

@bramathon that snippet doesn't:

* break out recursive calibrations
* capture which calibration was used to do the expansion

Some of the diff, also, improves how calibrations are handled and matched

**Comment by bramathon** [comment]
Yup, that's true. I don't have any need for those things for my use case.

File: quil-rs/src/program/mod.rs
Comment (kalzoo) [reply quil-rs/src/program/mod.rs]: I realize I had missed this comment on the last round, but it's a good idea. Changed that now.

**Comment by kalzoo** [comment]
Over to you now @bramathon

**Pull request merged**
